### ISSUE TITLE
RuntimeError: "dot" not implemented for 'Half'

### ISSUE THREAD

**Issue by NineMeowICT** [intel/intel-extension-for-pytorch#673]
### Describe the bug

Error message:

Traceback (most recent call last):
File "/media/ninemeow/Delta/LivePortrait/speed.py", line 209, in <module>
main()
File "/media/ninemeow/Delta/LivePortrait/speed.py", line 199, in main
warmupmodels(compiledmodels, stitchingretargeting_module, inputs)
File "/media/ninemeow/Delta/LivePortrait/speed.py", line 107, in warmupmodels
compiledmodels['SPADE Decoder'](inputs['generatorinput']) # Adjust input as required
File "/media/ninemeow/Delta/LivePortrait/venv/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1518, in wrappedcall_impl
return self.callimpl(args, *kwargs)
File "/media/ninemeow/Delta/LivePortrait/venv/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1527, in callimpl
return forward_call(args, *kwargs)
File "/media/ninemeow/Delta/LivePortrait/venv/lib/python3.9/site-packages/torch/dynamo/evalframe.py", line 328, in _fn
return fn(args, *kwargs)
File "/media/ninemeow/Delta/LivePortrait/venv/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1518, in wrappedcall_impl
return self.callimpl(args, *kwargs)
File "/media/ninemeow/Delta/LivePortrait/venv/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1527, in callimpl
return forward_call(args, *kwargs)
File "/media/ninemeow/Delta/LivePortrait/src/modules/spade_generator.py", line 44, in forward
x = self.Gmiddle0(x, seg)
File "/media/ninemeow/Delta/LivePortrait/venv/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1518, in wrappedcall_impl
return self.callimpl(args, *kwargs)
File "/media/ninemeow/Delta/LivePortrait/venv/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1527, in callimpl
return forward_call(args, *kwargs)
File "/media/ninemeow/Delta/LivePortrait/src/modules/util.py", line 303, in forward
dx = self.conv0(self.actvn(self.norm0(x, seg1)))
File "/media/ninemeow/Delta/LivePortrait/venv/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1518, in wrappedcall_impl
return self.callimpl(args, *kwargs)
File "/media/ninemeow/Delta/LivePortrait/venv/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1557, in callimpl
args_result = hook(self, args)
File "/media/ninemeow/Delta/LivePortrait/venv/lib/python3.9/site-packages/torch/nn/utils/spectralnorm.py", line 107, in _call__
setattr(module, self.name, self.computeweight(module, dopower_iteration=module.training))
File "/media/ninemeow/Delta/LivePortrait/venv/lib/python3.9/site-packages/torch/nn/utils/spectralnorm.py", line 93, in computeweight
sigma = torch.dot(u, torch.mv(weight_mat, v))
RuntimeError: "dot" not implemented for 'Half'

How to reproduce:
repo: https://github.com/NineMeowICT/LivePortrait-XPUOnly
command: python speed.py

### Versions

PyTorch version: 2.1.0.post2+cxx11.abi
PyTorch CXX11 ABI: Yes
IPEX version: 2.1.30+xpu
IPEX commit: 474a6b3cb
Build type: Release

OS: Linux Mint 21.3 (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
IGC version: 2024.1.0 (2024.1.0.20240308)
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.9.18 (main, Jul 11 2024, 09:10:56)  [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.5.0-41-generic-x86_64-with-glibc2.35
Is XPU available: True
DPCPP runtime version: 2024.1
MKL version: 2024.1
GPU models and configuration: 
[0] _DeviceProperties(name='Intel(R) Arc(TM) A770 Graphics', platform_name='Intel(R) Level-Zero', dev_type='gpu', driver_version='1.3.29138', has_fp64=0, total_memory=15473MB, max_compute_units=512, gpu_eu_count=512)
Intel OpenCL ICD version: 24.13.29138.29-881~22.04
Level Zero version: 1.3.29138.29-881~22.04

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      39 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             4
On-line CPU(s) list:                0-3
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Core(TM) i3-9100 CPU @ 3.60GHz
CPU family:                         6
Model:                              158
Thread(s) per core:                 1
Core(s) per socket:                 4
Socket(s):                          1
Stepping:                           11
CPU max MHz:                        4200.0000
CPU min MHz:                        800.0000
BogoMIPS:                           7200.00
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx rdseed adx smap clflushopt intel_pt xsaveopt xsavec xgetbv1 xsaves dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp vnmi md_clear flush_l1d arch_capabilities
Virtualization:                     VT-x
L1d cache:                          128 KiB (4 instances)
L1i cache:                          128 KiB (4 instances)
L2 cache:                           1 MiB (4 instances)
L3 cache:                           6 MiB (1 instance)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-3
Vulnerability Gather data sampling: Mitigation; Microcode
Vulnerability Itlb multihit:        KVM: Mitigation: VMX disabled
Vulnerability L1tf:                 Mitigation; PTE Inversion; VMX conditional cache flushes, SMT disabled
Vulnerability Mds:                  Mitigation; Clear CPU buffers; SMT disabled
Vulnerability Meltdown:             Mitigation; PTI
Vulnerability Mmio stale data:      Mitigation; Clear CPU buffers; SMT disabled
Vulnerability Retbleed:             Mitigation; IBRS
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; IBRS; IBPB conditional; STIBP disabled; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                Mitigation; Microcode
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] intel-extension-for-pytorch==2.1.30+xpu
[pip3] numpy==1.26.4
[pip3] torch==2.1.0.post2+cxx11.abi
[pip3] torchaudio==2.1.0.post2+cxx11.abi
[pip3] torchvision==0.16.0.post2+cxx11.abi
[conda] N/A

**Comment by wangkl2** [comment]
@NineMeowICT Thanks for reporting this issue. Will try to reproduce and get back to you later.

**Comment by wangkl2** [comment]
@NineMeowICT I can reproduce your issue. For now, could you please try AMP autocast first instead of half as a workaround?

Step1: Remove `.half()` method from the model of these two lines in your code: https://github.com/NineMeowICT/LivePortrait-XPUOnly/blob/main/speed.py#L82 and https://github.com/NineMeowICT/LivePortrait-XPUOnly/blob/main/speed.py#L89 

Step2: Enable autocast on XPU (either bf16 or fp16) during inference:
```
# Warm up models
with torch.xpu.amp.autocast(dtype=torch.float16):
    warm_up_models(compiled_models, stitching_retargeting_module, inputs)

# Measure inference times
with torch.xpu.amp.autocast(dtype=torch.float16):
    times, overall_times = measure_inference_times(compiled_models, stitching_retargeting_module, inputs)
```

I've tried either using FP32 or using AMP Autocast makes your workload work. And using autocast brings >2x performance boost than FP32 precision, especially for the "Warping Network" module and "SPADE Decoder" module. Please have a try.

For the issue of `RuntimeError: "dot" not implemented for 'Half'` inside Spectral_norm layer, I'll do some further debug and investigation.

**Comment by NineMeowICT** [comment]
@wangkl2 Thanks. I love oneAPI software stack and IPEX. Looking forward to your update.

**Comment by wangkl2** [comment]
@NineMeowICT Our developer has implemented the fp16 and bf16 dtype for `dot` op in this commit `https://github.com/intel/intel-extension-for-pytorch/commit/cb2e18192b847fd43075e6d319205df8781c3912`. 

And I've verified that it can solve the error of using `half` for the `spectral_norm` applying to the `conv` layer in your model code `https://github.com/NineMeowICT/LivePortrait-XPUOnly/blob/main/src/modules/util.py#L291`. The inference runs normally in either FP16 or BF16 now. This fix will be included in the IPEX release of `2.3.110+xpu`.

**Comment by NineMeowICT** [comment]
@wangkl2 Thank you. I will close the issue.
